### DISCUSSION TITLE
Option to hear perfect timed hitsounds when playing

### DISCUSSION THREAD

**Discussion by unholyghosttyper** [ppy/osu#19915] [Ideas]
Thoughts
As we all know, 0 audio offset means that hitsounds are perfectly in sync with bgm with intended rhythmic structure of the map.
But as a tradeoff we get inevitable audio latency on users end when you should hit your keys a bit earlier (up to 70ms in bad situations)
We can config our audio offset so it corresponds exactly to irl keypresses. 
But in that case hitsounds are shifted and no longer synced with music. So they become unreliable.
I don't know if it's possible to fix hitsounds offset latency with some additional latency on them.
But i think there's alternative that i would like to see in OSU!

Idea
Why not tradeoff being able to make hitsounds happen with being able to hear them playing perfectly when playing?
Can we have like separate audio track (kinda) with hitsounds always synced?
In all cases all players feel their keys when they press them and many of players actually listen to their keystrokes.

I mean... i have 25ms latency. 
With -25 offset i can perfectly rely on sync of my fingers and music. But there is no sync with hitsounds and music.  It's all over the place, so i play without hitsounds. 
Which means that in standard sliders can't give me feedback which can be NOT perfect. Sometimes it's really needed.
And in catch the beat playing without hitsounds is just cursed.

With 0 offiset i can perfectly rely on sync of music and hitsounds. But now there is another problem. 
You play from the past. Sometimes you play from reaaaly long time. 
So mania and taiko suffer most from it.
Yes, you can adjust to latency by playing a gigaton maps with it. But something is not right with that concept.

Hitsounds exist so you can hear if you are consistent with current rhythm and correct yourself accordingly if not.
If i have separate audio of perfectly played map synced with audio while i'm playin it i can correct myself same way if i was the source of hitsounds.
It doesn't give me any advantages. 
I still need to feel the rhythm the same way.
I still need to memorize the map.
I still need to be consistent.
It's just a way to sync music, mapped hitsounds and player.
It would've just be a reliable tool to see my mistakes and correct them.

And i mean, it should be only an option of course.

**Comment by peppy** [comment]
I don't foresee this being added. Hitsounds being triggered by user input is one of the core concepts behind osu!. I really dislike rhythm games that automate this playback.

We do have plans to add methods of reducing audio latency for those that are very sensitive to it, but of course this will require appropriate hardware to make work (low latency audio is hard and requires good driver and hardware support).

**Reply by unholyghosttyper** [reply]
I fully understand what you mean.
When game automates hitsounds with no other option its VERY annoying. A dance of Fire and ice does this.

It's good interaction with you and game when it works correctly.
Especially with custom hitsounded games like bemani games and djmax series.

I agree that osu in future should support low latency drivers like ASIO.
I tried different interfaces with different rhythm games. Right now i have TASCAM US-4x4.
As Osu don't support ASIO, latency is high.
Latest djmax does. 
And even then you will have latency.
Even with 192k samplerate and 32 buffer size you will experience like 5-6 ms latency.
Which is almost half long of perfect window of mania.
So if you play by hearing music it's gonna be late 300 every couple of notes. Instead of 300+.
And again, you must play early.
It works in djmax because it's window is like 50ms. So it's fine there.

So. I have dedicated audio interface.
And lets imagine the future where osu gets full ASIO support (or somethin' like that)
It gets 6ms of latency on my main PC with 192k samplerate and 32 buffer size, which should put my processor at work and make it really hot.
I have my ~1ms keyboard and ~1ms display.
I guess it's okay there. No complaints if it works perfectly. We'll get there.
8 ms is playable, i guess. When i record guitar it's fine.

But i like to play on my laptop when i'm not home.
It has display with some latency (~10ms)
It has generic realtek audio
It has membrane keyboard with 10ms latency
So in sum it has like ~50ms overall.
Which is horrendous.
Yeah i can take with me my cool keyboard and sound card (and display LUL)
But that's interesting expirience of sitting in cafe with all this.
Yeah, and tablet of course.

And like.... very little percentage of people will buy 300$ audio interface that have no issues with windows and drivers are stable.
(And still experience latency)
Even then it's not really convenient.
Kids can't buy it. It's not portable. It's not perfect.

Yeah, hearing perfect hitsounds as separate track is a quick rough not elegant at all fix to this problem.
But it solves too many problems at once with no repercussions than little time to adapt.
And it's not like hundreds of hours playing like when trying to adapt to 0 offset. It's like half an hour.
So, me and my friends would like to see even a glimpse of it.
We don't care if it's a mod, custom ruleset (if it's possible), ranked or unranked.

I don't know how hard it's to implement but i would strongly suggest team osu to at least consider it as an option.
Right now it's hitsounds with latency or no hitsounds at all for many of us.
I agree with you that hitsounds, generated by keypress is in a way fundamental to any rhythm game that have them.
And it feels better.
But we strongly believe it's the other way around too.
In every case in rhythm games you tweak yourself around the level. Not level tweaks around you.
High latency and inconsistent hitsounds breaks this principle apart, as you try to adapt not only to music but to latency AND hitsounds with said latency generated by you.
I think if it was a feature, nobody will lose anything but many people would win in that case.
Basically, we just want to play by music with right offset, but also want to hear hitsounds.
Which is impossible right now even if you give us low latency. It's still be latency.
Don't see it harming anyone.

If not, "I don't care about it so you shouldn't have too" is an honest take and we'll take it.
It's your game, not mine, not ours.

Thank you for quick reply and good day, sir.

**Reply by FeDuKOwO** [reply]
I agree with this guy. They have truth in their words and it would be cool if this feature was in our favourite game.

**Reply by shime27** [reply]
strongly agree with the op, it would make sense to do that

**Reply by didlgrave** [reply]
I'm VERY agree with this dude.

**Reply by Feodor0090** [reply]
There are no ways to fix bluetooth devices. All my friends who own only bt headphones usually just play without sound at all. Option to force hitsounds to be played synchronously will partially improve the situation here.

**Reply by ghost** [reply]
very, very late reply but i would be interested in implementing this myself.

**Reply by unholyghosttyper** [reply]
@loma-baldsson Still want this feature, looking forward to your implementation

**Reply by Alsen99** [reply]
I agree, this will help when we need to learn a rhythm, I believe maimai already done this by differentiating between the auto hitsound and real hitsound

**Reply by Alsen99** [reply]
And also will help with Auto Mod, since the hitsound will also be offsync because of global offset, you have to set the global offset to 0 to make the Auto Mod play in sync

**Reply by unholyghosttyper** [reply]
> And also will help with Auto Mod, since the hitsound will also be offsync because of global offset, you have to set the global offset to 0 to make the Auto Mod play in sync

I found a way to know your exact audio latency in game without relying on human execution.
I use offset to compensate it but since there is latency, hitsounds don't match.
I wouldn't mind, but i would like to hear at least sliderticks and tails.

Also if hitsounds are automated, it sounds like a cool feature to get around copyright when streaming. Maybe with metronome mod.

Plus some maps are quite mistimed. Would help if i can hear hitsounds.
And some are just mapped.... misteriously, so you just cannot know the rhythm unless you play auto

**Reply by ghost** [reply]
@unholyghosttyper i still don't really understand the proposition though. do you want it to play perfectly timed hitsounds even when you don't hit the hitobjectst? or do you want it to only play when you hit the hitobjects?

**Reply by unholyghosttyper** [reply]
> @unholyghosttyper i still don't really understand the proposition though. do you want it to play perfectly timed hitsounds even when you don't hit the hitobjectst? or do you want it to only play when you hit the hitobjects?
> 

It's not possible to play hitsounds when you hit objects in time when you use offset as there is always some latency and hitsounds are programmed to play as soon as possible. 
Which is at least 20ms for osu, for some people more.

So yeah, i just want them to play always

**Reply by KPoWasTaken** [reply]
Please reconsider.
I get that rgs forcing automated hitsounds with no way to turn it off can kinda suck. However, OP didn't ask for forced automation with no way to switch it off. OP asked for a toggle for automated hitsounds. OP is asking for an accessibility feature for those with high latency or those that can't really deal with any amount of latency. No matter what hardware, osu does have some amount of latency. Forcing people to choose between hitsounds with latency and no hitsounds but timed correctly with offset is kinda unfair imo and it's even more unfair on those with extreme latency as when it's too high, hitsounds with latency just becomes not an option. On top of that, many maps are designed around hitsounds. Unlike most other rgs, hitsounds in osu can be extremely important. So please reconsider this. There's no negatives to adding this. Those who don't want automated hitsounds would just leave it toggled off, and those who do deal with latency but still want hitsounds would toggle it on.

**Reply by Alsen99** [reply]
> Please reconsider.
> I get that rgs forcing automated hitsounds with no way to turn it off can kinda suck. However, OP didn't ask for forced automation with no way to switch it off. OP asked for a toggle for automated hitsounds. OP is asking for an accessibility feature for those with high latency or those that can't really deal with any amount of latency. No matter what hardware, osu does have some amount of latency. Forcing people to choose between hitsounds with latency and no hitsounds but timed correctly with offset is kinda unfair imo and it's even more unfair on those with extreme latency as when it's too high, hitsounds with latency just becomes not an option. On top of that, many maps are designed around hitsounds. Unlike most other rgs, hitsounds in osu can be extremely important. So please reconsider this. There's no negatives to adding this. Those who don't want automated hitsounds would just leave it toggled off, and those who do deal with latency but still want hitsounds would toggle it on.

Yes, please reconsider osu team. Being told by people to "just buy a new pc" is just inconsiderate of people who can't afford it.

**Comment by 1RainSeason** [comment]
Im agree with this man. Peppy lets go.

**Comment by Loreos7** [comment]
Just play with Relax mod if you feel and hear audio latency. Or buy a good laptop instead.

**Reply by unholyghosttyper** [reply]
I think you don't understand how latency works

**Reply by arily** [reply]
I sync my clicks to my keyboard noises

**Comment by abraker95** [comment]
I play with hitsounds muted. Adjusting audio offset and listening to cues in music is a much better experience for me. I tap to what I hear. and if I consistently hear sounds late, then it becomes a feedback loop where I tap later and later.

**Comment by NutchapolSal** [comment]
Please reconsider this. My setup has about a 200ms delay, which means I absolutely cannot play with hitsounds on or the audio will be an absolute mess. Implementing this would probably make osu! better than almost all other rhythm games out there as I've only seen this feature once but it makes so much difference.

PS. I found [a map](https://osu.ppy.sh/beatmapsets/302726#mania/678460) which have entire sections as hitsounds, which is cool but is completely ruined without having perfect timed hitsounds for me (Auto won't even let me enjoy the song as the hitsound delays are still there). 

PPS. "Qono Cafe" is that rhythm game, as the "Tap SE follows player" setting (where unchecking it makes hitsounds play on time)

**Reply by pigeonskew** [reply]
For starters,

1. Do not use wireless headphones.
2. Turn off any audio effect/enhancements you have currently enabled (like equalizers, etc).

**Comment by GalaxySpeaker** [comment]
Plz Plz Plz Plz i hope this can be used as an option
